.github/workflows/build.yml

```
@@ -0,0 +1,46 @@
+name: build
+
+on:
+  push:
+    paths-ignore:
+      - '*.rst'
+      - 'UNLICENCE'
+      - 'Doxyfile.in'
+      - 'shaders/**'
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        build_type: [Debug, Release]
+    
+    runs-on: ${{ matrix.os }}
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Install Ubuntu Packages
+      if: matrix.os == 'ubuntu-latest'
+      shell: bash
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y libassimp-dev libglfw3-dev
+
+    - name: Install MacOS Packages
+      if: matrix.os == 'macos-latest'
+      shell: bash
+      run: |
+        brew install assimp
+        brew install glfw
+
+    - name: Generate
+      shell: bash
+      run: |
+        [[ ${{ matrix.os }} = 'macos-latest' ]] && generator='Xcode' || generator='Ninja'
+        cmake . -G"${generator}" -B${{ github.workspace }}/build -DCMAKE_BUILD_TYPE=${{ matrix.build_type }}
```

```suggestion
        cmake -G"${generator}" -S${{ github.workspace }} -B${{ github.workspace }}/build -DCMAKE_BUILD_TYPE=${{ matrix.build_type }}
```
1. Using `-S` to explicitly set the source folder: when using '.' as the source, it is easier to spot if prefixed by `-S` compared to standing on its own, and it’s also more explicit about what that path is.
2. Similar to using `${{ github.workspace }}` instead of '.' for the build folder, but applying it to the source folder.

---

Good catch, I didn't notice that

---

It's still complaining about Ninja though, I'm not entirely sure why, maybe you have any ideas?

---

Yes: the Ubuntu environment does not seem to come with Ninja, so the _ninja-build_ package needs to be installed as well.

For Windows, Ninja is only exposed when using the Visual Studio command prompt which sets everything right for picking the appropriate C/C++ compiler, etc. Trying to figure out how to get access to it in the environment.

---

So, for Windows the recommended solution seems to be to use the following action to get access to the dev command prompt: https://github.com/marketplace/actions/enable-developer-command-prompt.

However there are some issues when [using bash with it](https://github.com/marketplace/actions/enable-developer-command-prompt#name-conflicts-with-shell-bash), so using a different shell there would be better. The easiest would be to remove all the `shell: bash` found in the config file and use the default shell (which will be `bash` on non-Windows platform, and `pwsh` on Windows. If doing that, the “Generate” step should be duplicated as `[[ ${{ matrix.os }} = 'macos-latest' ]] && generator='Xcode' || generator='Ninja'` will not work with PowerShell.

---

I think removing the `shell: bash` might also be needed for something else: `${{ github.workspace }}` expands to a Windows-style path (with `\` as separators) on Windows, and [bash is not fond of them](https://github.com/Husenap/CG_Labs/runs/3405007075?check_suite_focus=true#step:5:3) (from the latest actions run).

---

Good points, I've made some changes and it looks promising. I managed to workaround the generator problem by moving it from the shell to Github Actions' env variables, so I can use their expression syntax to do the same thing. 

https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#operators
https://github.community/t/do-expressions-support-ternary-operators-to-change-their-returned-value/18114/3

---

Last build is successful :) I'll look into it more tomorrow!

---

Ah, well done for working around the generator issue and getting everything building again!